### ISSUE TITLE
Completing login after signInRedirect

### ISSUE THREAD

**Issue by elfico** [authts/react-oidc-context#99]
Hi,

We are trying to switch to this library instead of vanilla `oidc-client`.

In our previous implementation, after calling `signInRedirect `when the authentication is returned, our `redirect_url` page contains code that will complete the login by calling `userManager.signinRedirectCallback()` which completes the authentication by exchanging the authorization code for an access token.

I can't find an example of how this is done in the library.

I would also like to know if this library handles `silentRenewal` and how to implement.

**Comment by elfico** [comment]
Thanks for the speedy response @pamapa . I have been able to get it working.

**Comment by edwardsilhol** [comment]
Hi guys and thank you so much for building and open sourcing this very useful library ! You rock.

Could I suggest a couple of really small adjustments to the readme that would make it easier for people to understand how to use it ?

1. Mention that the config object can take more options than the oAuth conf (and perhaps list them and their roles). Somehow I didn't find that in the code for a while and lost quite a lot of time as a result trying to implement some of the callbacks myself.
2. Add an example combining the persisted sessionStorage token and silentSignin method. Persisting the login state upon page reload seems like a very natural thing to do, I was surprised it wasn't done by the library by default (though I'm sure you had a good reason not to do it). I was able to make my own implementation but it took some time getting familiar with the library and I reckon other newcomers would appreciate an example. Happy to share my code if useful at all.

Thanks again, all the best !!

**Comment by pamapa** [comment]
@edwardsilhol Yes please share your code here.

**Comment by edwardsilhol** [comment]
This is what I ended up doing (not thoroughly tested yet) :

```
import { useEffect } from 'react';
import { useAuth } from 'react-oidc-context';

import { CLIENT_ID, OAUTH2_AUTHORITY } from '@/utils/constants';

const useAuthSession = () => {
  const auth = useAuth();

  useEffect(() => {
    const checkIfExistingSession = async () => {
      const storedState = sessionStorage.getItem(
        `oidc.user:${OAUTH2_AUTHORITY}:${CLIENT_ID}`,
      );

      if (!auth.isLoading && !auth.isAuthenticated && storedState) {
        await auth.signinSilent();
      }
    };

    void checkIfExistingSession();
  }, [auth]);

  return auth;
};
```

**Comment by pamapa** [comment]
Are you aware of the [automaticSilentRenew](https://authts.github.io/oidc-client-ts/interfaces/UserManagerSettings.html#automaticSilentRenew) feature of `oidc-client-ts`? With that you will not need to handle the renew your self...

**Comment by edwardsilhol** [comment]
Hm, I think this worries only about refreshing the token before it expires, and they say it's true by default (unless disabled by your lib?) .

Following the instructions in your Readme the login state wasn't restored upon page reload.

**Comment by pamapa** [comment]
A page reload executes [`userManager.getUser()`](https://github.com/authts/react-oidc-context/blob/f175dcba6ab09871b027d6a2f2224a17712b67c5/src/AuthProvider.tsx#L160), which triggers all needed events including access token expired, which should lead to a automatic silent renew (if still possible)...

**Comment by edwardsilhol** [comment]
Hi, so I figured out why it wasn't working, posting here in case someone else runs into this : 

You must provide an implementation of onSigninCallback to the conf to remove the parameters from the URL upon successful login. Otherwise if you refresh the page and the params are still there the silent login won't work.

I know an example is provided [in your code](https://github.com/authts/react-oidc-context/blob/f175dcba6ab09871b027d6a2f2224a17712b67c5/src/AuthProvider.tsx#L20), but it would be really useful to add it to the readme too, it took me a while to find it.

For anyone reading this and having the same issue : 

const onSigninCallback = (): void => {
  window.history.replaceState({}, document.title, window.location.pathname);
};

With this you don't need the effect I posted above.

Best
